lib/scholar/interpolation/cubic_spline.ex

```diff
@@ -0,0 +1,189 @@
+defmodule Scholar.Interpolation.CubicSpline do
+  @moduledoc """
+  Cubic spline interpolation
+  """
+  import Nx.Defn
+
+  @derive {Nx.Container, containers: [:coefficients, :x]}
+  defstruct [:coefficients, :x]
+
+  opts = [
+    extrapolate: [
+      required: false,
+      default: true,
+      type: :boolean,
+      doc: "if false, out-of-bounds x return NaN."
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+
+  @doc """
+  Fits a cubic spline interpolation of the given `(x, y)` points
+
+  Inputs are expected to be rank-1 tensors with the same shape
+  and at least 3 entries.
+  """
+  defn train(x, y) do
+    # https://en.wikiversity.org/wiki/Cubic_Spline_Interpolation
+    # Reference implementation in Scipy
+
+    {n, dx} =
+      case Nx.shape(x) do
+        {n} when n > 2 ->
+          dx = x[1..-1//1] - x[0..-2//1]
+          {n, dx}
+
+        shape ->
+          raise ArgumentError,
+                "expected x to be a tensor with shape {n}, where n > 2, got: #{inspect(shape)}"
+      end
+
+    case {Nx.shape(y), Nx.shape(x)} do
+      {shape, shape} ->
+        :ok
+
+      {y_shape, x_shape} ->
+        raise ArgumentError,
+              "expected y to have shape #{inspect(x_shape)}, got: #{inspect(y_shape)}"
+    end
+
+    sort_idx = Nx.argsort(x)
+    x = Nx.take_along_axis(x, sort_idx)
+    y = Nx.take_along_axis(y, sort_idx)
```

Comment:
```suggestion
    x = Nx.take(x, sort_idx)
    y = Nx.take(y, sort_idx)
```

For 1D tensor it should be the same and I think take is less involved as it always uses the indices as is (at least in exla).

Reply:
For torchx it's the opposite haha! They provide a gather function which is exactly our take_along_axis

Reply:
Oh well :see_no_evil: Good to know!